[Discord Notification wrapper]

Tytouc opened the pull request:
as per issue #599

**Overview**
Sample script to wrap raspiBackup.sh in order to send Discord notification message once raspiBackup is finished
_Tested with raspiBackup.sh version: 0.6.8 CommitSHA: 120287b CommitDate: 2022-12-15 CommitTime: 16:48:38_

Relies on raspiBackup variables:
  `BACKUP_TARGETDIR` refers to the backupdirectory just created. Used when raspiBackup succeeded to read the raspiBackup.msg file content
In case raspiBakup failed, the file raspiBackup.msg file under HOME folder of calling user is used


**Requires:**

- [curl ](https://curl.se):
		for sending POST command to Discord WebHook
		_Tested with version: curl 7.64.0_

- 	[jq ](https://stedolan.github.io/jq/):
		For JSON manipulation in shell
		_Tested with version jq-1.5-1-a5b5cbe_

**Configuration:**
 	-> file: `/usr/local/etc/discordWrapper.conf`
 	-> As it contains an URL that allows anyone to send message to Discord Channel, this file **shall have `0600 root:root` permissions only (-rw-r-----)**
	-> Content of the config file:
```
# URL of the Discord WebHook where to send the message. Adapt to your case
CONFIG_DISCORD_URL="https://discord.com/api/webhooks/1234567890.../abcdef...."
# Possible Colors to be used for Discord Embed message. Must be integer value in base 10 for RGB clolor coding. Refer to [Discord documentation](https://discord.com/developers/docs/intro)
COLOR_BLUE=255
COLOR_RED=16711680
COLOR_GREEN=65280

 # Color to be used depending on the raspiBackup execution result: success or failure
 # each line shall refer to one of the color defined here upper
CONFIG_COLOR_DEFAULT=$COLOR_BLUE
CONFIG_COLOR_SUCCESS=$COLOR_GREEN
CONFIG_COLOR_FAILURE=$COLOR_RED`
```

Tytouc committed:
Addint Discord notification wrapper script and configuration file

as per #559 issue created

framps commented on extensions_userprovided/raspiBackupDiscordWrapper.sh:
```diff
@@ -0,0 +1,241 @@
+
+set -euf -o pipefail
+
+MYSELF=${0##*/}
```
The whole git handling stuff will work only if the code will be published in a special way I use to publish raspiBackup parts. You can keep the code - but you ma just use `VERSION`

Tytouc replied:
should I remove `set -euf -o pipefail` ?

Tytouc replied:
done

framps commented on extensions_userprovided/raspiBackupDiscordWrapper.sh:
```diff
@@ -0,0 +1,241 @@
+#### DISCORD SECTION
+
+# read config file
+source /usr/local/etc/discordWrapper.conf
```
I think a comment to copy and update the conf file discordWrapper.conf into /usr/local/etc would be helpful.

framps commented on extensions_userprovided/raspiBackupDiscordWrapper.sh:
```diff
@@ -0,0 +1,241 @@
+###########################################
+### Default JSON message for Discord notification with a colored embed
+###########################################
+DEFAULTJSON=$(cat<<EOS
```
Nice way to initialize a bash variable with a here doc. I prefer to use read (see [this page](https://stackoverflow.com/questions/1167746/how-to-assign-a-heredoc-value-to-a-variable-in-bash)) :-)

Tytouc replied:
This is to preserve formatting

framps commented on extensions_userprovided/raspiBackupDiscordWrapper.sh:
```diff
@@ -0,0 +1,241 @@
+# these variables are set in function send_discord
+###########################################
+function build_json() {
+        DISCORD_JSON=$(jq \
```
I you run shellcheck you will get warning because $(...) is not quoted. No issue but I use quotes as much as possible to avoid any splitting of strings done by bash

Tytouc replied:
I see

Tytouc replied:
added "" around $(...)

framps commented on extensions_userprovided/raspiBackupDiscordWrapper.sh:
```diff
@@ -0,0 +1,241 @@
+###########################################
+function send_discord() { # $1:status of the operation, 0=) success, not 0 is failure
+        local discord_color=$CONFIG_COLOR_DEFAULT
+        if [[ $1 -ne  0 ]]; then
```
I prefer to use `(( $1 != 0 ))` which is more readable

Tytouc replied:
Correct, we all have some habits ;)

Tytouc replied:
done

framps commented on extensions_userprovided/raspiBackupDiscordWrapper.sh:
```diff
@@ -0,0 +1,241 @@
+        fi
+        DISC_CONTENT="Raspi Backup Status: $(date)"
+        DISC_TITLE="Raspi Backup $TITLE_SUFFIX"
+	# just to ensure not to fail in case the message file is not found and send a minimal status message without details
```
:+1:

framps commented on extensions_userprovided/raspiBackupDiscordWrapper.sh:
```diff
@@ -0,0 +1,241 @@
+
+        DISCORD_URL=$CONFIG_DISCORD_URL
+
+        curl -s -H "Content-Type: application/json" -X POST -d "$DISCORD_JSON" $DISCORD_URL > /dev/null
```
Is there any reason you suppress any error messages?

Tytouc replied:
curl is sending 'progress' and other stuff on console, which I wanted to avoid. If curl fails, this is not a big deal.

framps commented on extensions_userprovided/raspiBackupDiscordWrapper.sh:
```diff
@@ -0,0 +1,241 @@
+
+# main program
+
+if [[ -z $CONFIG_DISCORD_URL ]]; then
```
:+1:

framps commented on extensions_userprovided/raspiBackupDiscordWrapper.sh:
```diff
@@ -0,0 +1,241 @@
+# user currently calling the script, if called with sudo it gives the sudoer user name
+USER=$(logname)
+# home folder of the user
+MSG_FILE_DIR="$(eval echo "~$USER")"
```
:+1:

Tytouc committed:
Update raspiBackupDiscordWrapper.sh

removed un-necessary GIT stuff etc
changed way of writing from [[ $1 -ne  0 ]] to (( $1 != 0 ))

Tytouc committed:
added quote around $(...)

in function build_json()

Tytouc committed:
added config file comments
